### PR TITLE
KM96: Add command prompt (PS1) for bash and zsh

### PR DESCRIPTION
**Pull request by yngvark**
Set PS1 (i.e. a custom command prompt) for the sub shell run by `okctl venv`.

## Description
When running `okctl venv myenv`, a sub shell is run. This PR sets the PS1 environment variable to show

```
<current dir> (<okctl environment:k8s namespace>)
```

See screenshot below.

More tests and documentation will be added in later PRs if this PR is accepted.

## Motivation and Context

* User feedback from demo stated that it would be nice to see some change when running `okctl venv`.
* We want the user not to run commands accidentally in the wrong k8s cluster, if he uses an existing k8s cluster already.

kube-ps1 does the same thing as already. However, to make using okctl as simple as possible, we come with this functionally pre-shipped.

The user may also want to use kube-ps1 instead. For this case, I'll make a new PR that adds the flag `--no-ps1` and an environment variable or something similar to `okctl venv`.

Options
* The user may want to use a custom PS1. In this case, s(he) can set OKCTL_PS1=your-own-ps1.
* The user  may want to keep her existing PS1. This is possible by setting OKCTL_NO_PS1=true

## How Has This Been Tested?

I wasn't sure if this PR would be accepted the way it is (the zsh hacks may be to hackish too accept). If it is accepted, I will make a new PR with more tests. I just didn't want to waste time.

I have tested this manually.

**BUG found** by @deifyed , __git_ps1 is not found when running in zsh. I'll fix that.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (See top comment about more PRs)
- [x] I have added tests to cover my changes. (Somewhat, see comment under test)
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by deifyed** [comment]
Got this when running `okctl venv test` (in addition to the expected output)

```sh
prompt:1: command not found: __git_ps1
~/origo/kjøremiljø/okctl-iac  test:default $
```

**Review by deifyed — APPROVED** [review]
Looks good to my untrained go eyes

File: cmd/okctl/venv.go [hunk 1]
```diff
@@ -97,27 +135,100 @@ func buildVenvCommand(o *okctl.Okctl) *cobra.Command {
 	return cmd
 }
 
-func printExecutables(env []string) error {
-	whichKubectl, err := getWhich("kubectl", env)
+func setCmdPromptZsh(opts *virtualenv.VirtualEnvironmentOpts, venv *virtualenv.VirtualEnvironment) (*storage.TemporaryStorage, error) {
+	userHomeDir, err := os.UserHomeDir()
+	if err != nil {
+		return nil, fmt.Errorf("couldn't set command prompt: %w", err)
```
[the remark this answers is not in the document]
Comment (yngvark) [reply cmd/okctl/venv.go:141]: Yeah, that makes sense, will fix.

File: cmd/okctl/venv.go [hunk 2]
```diff
@@ -65,25 +80,48 @@ func buildVenvCommand(o *okctl.Okctl) *cobra.Command {
...
+			case virtualenv.ShellIsBash(shellCmd):
+				virtualenv.SetCmdPromptBash(&opts, venv)
+			case virtualenv.ShellIsZsh(shellCmd):
+				zshrcTmpWriteStorage, err := setCmdPromptZsh(&opts, venv)
+				if err != nil {
+					return err
+				}
+
+				defer func() {
+					// This leaves an empty directory /tmp/okctl538407010. We should clean this up.
```
[the remark this answers is not in the document]
Comment (yngvark) [reply cmd/okctl/venv.go:103]: Right, removing comment.

**Review by paulbes — COMMENTED** [review]
I haven't verified this very deeply, but the code below is an attempt at restructuring the virtualenv package. Golang relies heavily on composition, embedding and dependency injection to achieve scalable and reusable code. 

```go
// ShellType enumerates shells we recognize
type ShellType string

// nolint: golint
const (
	ShellTypeBash ShellType = "bash"
	ShellTypeZsh  ShellType = "zsh"
)

// ShellGetter provides an interface for interacting with
// the shell of a user
// We can then provide concrete implementations of the shell getter:
// We can do so by switching on the GOOS variable, for example, seeing
// how different versions of an operating system will have different
// mechanisms for retrieval.
// We can also have a static getter for testing, for example.
type ShellGetter interface {
	// Get returns the shell
	Get() (ShellType, error)
}

// CommandLinePrompter provides an interface for
// configuring the command line prompt
type CommandLinePrompter interface {
	// BuildPrompt
	BuildPrompt(opts VirtualEnvironmentOpts) ([]byte, error)
	SetPrompt([]byte) error
	Cleanup() error
}

type VirtualEnvCreator interface {
	Create(opts VirtualEnvironmentOpts) error
}

type VirtualEnvCleaner interface {
	Clean() error
}

type VirtualEnver interface {
	VirtualEnvCreator
	VirtualEnvCleaner
}

// VirtualEnv contains state for maintaining a
type VirtualEnv struct {
	shell  ShellGetter // Maybe this shouldn't be stored in here, but rather jsut used to create the proer prompter
	prompt CommandLinePrompter
}

// Ensure that VirtualEnv implementer the VirtualEnver interface
var _ VirtualEnver = &VirtualEnv{}

// Create the virtual env by orchestrating all the different parts
func (v *VirtualEnv) Create(opts VirtualEnvironmentOpts) error {
	t, err := v.shell.Get()
	if err != nil {
		return err
	}

	switch t {
	case ShellTypeBash:
		// create a bash prompter
	case ShellTypeZsh:
		// create a zsh prompter
	default:
		return fmt.Errorf("unknown shell type: %s", t)
	}

	data, err := v.prompt.BuildPrompt(opts)
	if err != nil {
		return err
	}

	err = v.prompt.SetPrompt(data)
	if err != nil {
		return err
	}
	
	return nil
}

// Clean removes all remnants, if any such exist
func (v *VirtualEnv) Clean() error {
	return v.prompt.Cleanup()
}

// New returns an initialised shell interactor
func New(shell ShellGetter) *VirtualEnv {
	return &VirtualEnv{
		shell: shell,
	}
}
```

**Comment by yngvark** [comment]
Closing, replaced by https://github.com/oslokommune/okctl/pull/144

**Pull request closed without merging**